### ISSUE TITLE
Cannot build for WASM target

### ISSUE THREAD

**Issue by kylejlin** [koka-lang/koka#579]
I tried compiling the `fibonacci.kk` example with the WASM target, but got an error.

## Expected behavior:

The compiler emits a `.wasm` file.

## Actual behavior:

The compiler emits an error. I copied and pasted my terminal's contents below.

```txt
(base) kyle@Kyles-MBP-7 koka_playground % koka fibonacci.kk --target=wasm --cc=/Users/kyle/emsdk/upstream/emscripten/emcc
load    : std/core
load    : std/core/types
load    : std/core/unsafe
load    : std/core/vector
load    : std/core/exn
load    : std/core/bool
load    : std/core/undiv
load    : std/core/hnd
load    : std/core/char
load    : std/core/sslice
load    : std/core/string
load    : std/core/order
load    : std/core/int
load    : std/core/either
load    : std/core/list
load    : std/core/maybe
load    : std/core/tuple
load    : std/core/show
load    : std/core/delayed
load    : std/core/console
load    : std/core/debug
parse   : .../fibonacci.kk
check   : fibonacci
linking : fibonacci/@main
compile : kklib from: /usr/local/share/koka/v3.1.2/kklib
fatal error: error in backend: Cannot select: intrinsic %llvm.thread.pointer
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /Users/kyle/emsdk/upstream/bin/clang -target wasm32-unknown-emscripten -fignore-exceptions -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --sysroot=/Users/kyle/emsdk/upstream/emscripten/cache/sysroot -DEMSCRIPTEN -Werror=implicit-function-declaration -Xclang -iwithsysroot/include/fakesdl -Xclang -iwithsysroot/include/compat -Wall -Wextra -Wpointer-arith -Wshadow -Wstrict-aliasing -Wno-unknown-pragmas -Wno-missing-field-initializers -Wno-unused-parameter -Wno-unused-variable -Wno-unused-value -Wno-unused-but-set-variable -g3 -Og -c -D__wasi__ -I/usr/local/share/koka/v3.1.2/kklib/include -I/usr/local/share/koka/v3.1.2/kklib/mimalloc/include -DKK_MIMALLOC=4 -DKK_COMP_VERSION=\"3.1.2\" -DKK_CC_NAME=\"emcc\" /usr/local/share/koka/v3.1.2/kklib/src/all.c -o .koka/v3.1.2/emcc-wasm32-debug-68cf4c/kklib.o
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module '/usr/local/share/koka/v3.1.2/kklib/src/all.c'.
4.	Running pass 'WebAssembly Instruction Selection' on function '@mi_heap_malloc_small_zero'
 #0 0x000000010573303c llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/Users/kyle/emsdk/upstream/bin/clang-19+0x10113703c)
 #1 0x0000000105730e94 llvm::sys::RunSignalHandlers() (/Users/kyle/emsdk/upstream/bin/clang-19+0x101134e94)
 #2 0x0000000105732598 llvm::sys::CleanupOnSignal(unsigned long) (/Users/kyle/emsdk/upstream/bin/clang-19+0x101136598)
 #3 0x00000001056a27d4 (anonymous namespace)::CrashRecoveryContextImpl::HandleCrash(int, unsigned long) (.llvm.18080130669317311330) (/Users/kyle/emsdk/upstream/bin/clang-19+0x1010a67d4)
 #4 0x00000001056a2780 llvm::CrashRecoveryContext::HandleExit(int) (/Users/kyle/emsdk/upstream/bin/clang-19+0x1010a6780)
 #5 0x000000010572dcc4 llvm::sys::Process::ExitNoCleanup(int) (/Users/kyle/emsdk/upstream/bin/clang-19+0x101131cc4)
 #6 0x0000000104686938 std::__2::__function::__policy_invoker<bool (llvm::Function const&)>::__call_empty[abi:nn190000](std::__2::__function::__policy_storage const*, llvm::Function const&) (/Users/kyle/emsdk/upstream/bin/clang-19+0x10008a938)
 #7 0x00000001056a76b4 llvm::report_fatal_error(llvm::Twine const&, bool) (/Users/kyle/emsdk/upstream/bin/clang-19+0x1010ab6b4)
 #8 0x0000000106797700 llvm::SelectionDAGISel::CannotYetSelect(llvm::SDNode*) (/Users/kyle/emsdk/upstream/bin/clang-19+0x10219b700)
 #9 0x0000000106796d24 llvm::SelectionDAGISel::SelectCodeCommon(llvm::SDNode*, unsigned char const*, unsigned int) (/Users/kyle/emsdk/upstream/bin/clang-19+0x10219ad24)
#10 0x00000001046e83c8 (anonymous namespace)::WebAssemblyDAGToDAGISel::Select(llvm::SDNode*) (.llvm.2764733163389807524) (/Users/kyle/emsdk/upstream/bin/clang-19+0x1000ec3c8)
#11 0x000000010678f1f0 llvm::SelectionDAGISel::DoInstructionSelection() (/Users/kyle/emsdk/upstream/bin/clang-19+0x1021931f0)
#12 0x000000010678e930 llvm::SelectionDAGISel::CodeGenAndEmitDAG() (/Users/kyle/emsdk/upstream/bin/clang-19+0x102192930)
#13 0x000000010678dcb8 llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) (/Users/kyle/emsdk/upstream/bin/clang-19+0x102191cb8)
#14 0x000000010678b354 llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) (/Users/kyle/emsdk/upstream/bin/clang-19+0x10218f354)
#15 0x0000000104db5e6c llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/Users/kyle/emsdk/upstream/bin/clang-19+0x1007b9e6c)
#16 0x00000001052841cc llvm::FPPassManager::runOnFunction(llvm::Function&) (/Users/kyle/emsdk/upstream/bin/clang-19+0x100c881cc)
#17 0x000000010528cc0c llvm::FPPassManager::runOnModule(llvm::Module&) (/Users/kyle/emsdk/upstream/bin/clang-19+0x100c90c0c)
#18 0x0000000105284c98 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/Users/kyle/emsdk/upstream/bin/clang-19+0x100c88c98)
#19 0x0000000105ef347c clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::__2::unique_ptr<llvm::raw_pwrite_stream, std::__2::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) (/Users/kyle/emsdk/upstream/bin/clang-19+0x1018f747c)
#20 0x0000000105d952a4 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/Users/kyle/emsdk/upstream/bin/clang-19+0x1017992a4)
#21 0x0000000106f192a8 clang::ParseAST(clang::Sema&, bool, bool) (/Users/kyle/emsdk/upstream/bin/clang-19+0x10291d2a8)
#22 0x00000001061f950c clang::FrontendAction::Execute() (/Users/kyle/emsdk/upstream/bin/clang-19+0x101bfd50c)
#23 0x000000010616f478 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/Users/kyle/emsdk/upstream/bin/clang-19+0x101b73478)
#24 0x00000001062d51e0 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/Users/kyle/emsdk/upstream/bin/clang-19+0x101cd91e0)
#25 0x000000010468603c cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/Users/kyle/emsdk/upstream/bin/clang-19+0x10008a03c)
#26 0x0000000104683b80 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) (/Users/kyle/emsdk/upstream/bin/clang-19+0x100087b80)
#27 0x000000010600e750 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::__2::optional<llvm::StringRef>>, std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char>>*, bool*) const::$_0>(long) (/Users/kyle/emsdk/upstream/bin/clang-19+0x101a12750)
#28 0x00000001056a2758 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/Users/kyle/emsdk/upstream/bin/clang-19+0x1010a6758)
#29 0x000000010600e17c clang::driver::CC1Command::Execute(llvm::ArrayRef<std::__2::optional<llvm::StringRef>>, std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char>>*, bool*) const (/Users/kyle/emsdk/upstream/bin/clang-19+0x101a1217c)
#30 0x0000000105fd425c clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/Users/kyle/emsdk/upstream/bin/clang-19+0x1019d825c)
#31 0x0000000105fd44c4 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::__2::pair<int, clang::driver::Command const*>>&, bool) const (/Users/kyle/emsdk/upstream/bin/clang-19+0x1019d84c4)
#32 0x0000000105ff0f24 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::__2::pair<int, clang::driver::Command const*>>&) (/Users/kyle/emsdk/upstream/bin/clang-19+0x1019f4f24)
#33 0x0000000104682bcc clang_main(int, char**, llvm::ToolContext const&) (/Users/kyle/emsdk/upstream/bin/clang-19+0x100086bcc)
#34 0x00000001046904fc main (/Users/kyle/emsdk/upstream/bin/clang-19+0x1000944fc)
#35 0x0000000185ae50e0 
clang: error: clang frontend command failed with exit code 70 (use -v to see invocation)
clang version 19.0.0git (https:/github.com/llvm/llvm-project bc9823cf60bf91cc8b45248c4205cd2c67b2a3d5)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: /Users/kyle/emsdk/upstream/bin
clang: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /var/folders/l1/mdzf4c61045_w21mwjrnpls80000gn/T/all-11755c.c
clang: note: diagnostic msg: /var/folders/l1/mdzf4c61045_w21mwjrnpls80000gn/T/all-11755c.sh
clang: note: diagnostic msg: Crash backtrace is located in
clang: note: diagnostic msg: /Users/kyle/Library/Logs/DiagnosticReports/clang-19_<YYYY-MM-DD-HHMMSS>_<hostname>.crash
clang: note: diagnostic msg: (choose the .crash file that corresponds to your crash)
clang: note: diagnostic msg: 

********************
fibonacci/@main(1, 1): build error: user error (error  : command failed (exit code 1)
command: /Users/kyle/emsdk/upstream/emscripten/emcc -Wall -Wextra -Wpointer-arith -Wshadow -Wstrict-aliasing -Wno-unknown-pragmas -Wno-missing-field-initializers -Wno-unused-parameter -Wno-unused-variable -Wno-unused-value -Wno-unused-but-set-variable -g -Og -c -D__wasi__ -I /usr/local/share/koka/v3.1.2/kklib/include -I /usr/local/share/koka/v3.1.2/kklib/mimalloc/include -DKK_MIMALLOC=4 "-DKK_COMP_VERSION=\"3.1.2\"" "-DKK_CC_NAME=\"emcc\"" -o .koka/v3.1.2/emcc-wasm32-debug-68cf4c/kklib.o /usr/local/share/koka/v3.1.2/kklib/src/all.c)

Failed to compile fibonacci.kk
(base) kyle@Kyles-MBP-7 koka_playground % 
```

## Debug and version info

**`koka`**

```txt
(base) kyle@Kyles-MBP-7 koka_playground % koka --version
Koka 3.1.2, 17:54:10 May 30 2024 (ghc release version)
version: 3.1.2
bin    : /usr/local/bin
lib    : /usr/local/lib/koka/v3.1.2
share  : /usr/local/share/koka/v3.1.2
output : .koka/v3.1.2/clang-debug-319c39
cc     : /usr/bin/clang
```

**`emcc` (Emscripten):**

```txt
(base) kyle@Kyles-MBP-7 koka_playground % /Users/kyle/emsdk/upstream/emscripten/emcc --version                           
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.60 (42a6ea2052f19f70d7d994e8c324bcad2f1f8939)
```

**Emscripten-vendored `clang`:**

```txt
(base) kyle@Kyles-MBP-7 koka_playground % /Users/kyle/emsdk/upstream/bin/clang --version
clang version 19.0.0git (https:/github.com/llvm/llvm-project bc9823cf60bf91cc8b45248c4205cd2c67b2a3d5)
Target: x86_64-apple-darwin23.3.0
Thread model: posix
```

**`fibonacci.kk`:**

```koka
// Fibonacci numbers.
// The `:int` type in Koka has infinite precision.
module fibonacci

pub fun main()
  val n = 10000
  println("The " ++ n.show ++ "th fibonacci number is " ++ fibonacci(10000).show)

// Calculate the `n`'th fibonacci number.
fun fibonacci(n : int) : div int
  fib(n, 0, 1)


fun fib(n : int, x1 : int, x2 : int) : div int
  if n<=0 then x1 else fib(n - 1, x2, x1+x2)

pub fun examplex()
  fibonacci(100).println

fun example2()
  fibonacci(1000).println
```

**Device:**
- Model: Macbook Pro 2021
- Chip: Apple M1 Pro
- OS: macOS Sonoma 14.3.1

## Steps to reproduce

1. Copy the `fibonacci.kk` file provided above.
2. Install koka 3.1.2
3. Install Emscripten 3.1.60. If needed, activate version 3.1.60 after installing it.
4. Run `koka fibonacci.kk --target=wasm --cc=/PATH/TO/YOUR/EMSCRIPTEN/emcc`

## Personal thoughts

It is possible that this is not a Koka compiler bug, but rather an Emscripten bug (or a bug with one of Emscripten's dependencies, such as LLVM clang). In any case, given that I installed both the latest Koka release and the latest Emscripten release, this suggests that **the latest Koka release is incompatible with the latest Emscripten release.** As far as I can tell, the current docs do not make note of any unsupported Emscripten versions, so I suggest updating the docs with the range of supported Emscripten versions.

**Comment by daanx** [comment]
Mm, it looks like it tries to get the `builtin_thread_pointer` which is not supported on WASM. Maybe Koka doesn't pass the right flag to mimalloc. Can you try to run Koka with `-fstdalloc` and see if you still get the same issue?

(also, the clang is for x86_64-apple-darwin23.3.0 which means it runs in emulation on the M1 ? but maybe that is how it is set up for emscripten, not sure)

**Comment by kylejlin** [comment]
I tried adding the `--fstdalloc` flag. The old error seems fixed, but I got new errors. Here is the full output:

```txt
(base) kyle@Kyles-MBP-7 koka_playground % koka fibonacci.kk --target=wasm --cc=/Users/kyle/emsdk/upstream/emscripten/emcc --fstdalloc
load    : std/core
load    : std/core/undiv
load    : std/core/vector
load    : std/core/unsafe
load    : std/core/hnd
load    : std/core/exn
load    : std/core/bool
load    : std/core/order
load    : std/core/int
load    : std/core/string
load    : std/core/list
load    : std/core/sslice
load    : std/core/maybe
load    : std/core/either
load    : std/core/tuple
load    : std/core/char
load    : std/core/show
load    : std/core/debug
load    : std/core/delayed
load    : std/core/console
parse   : .../std/core/types.kk
parse   : .../std/core/undiv.kk
parse   : .../std/core/unsafe.kk
parse   : .../std/core/hnd.kk
parse   : .../std/core/int.kk
parse   : .../std/core/order.kk
parse   : .../std/core/exn.kk
parse   : .../std/core/bool.kk
parse   : .../std/core/vector.kk
parse   : .../std/core/char.kk
parse   : .../std/core/list.kk
parse   : .../std/core/sslice.kk
parse   : .../std/core/string.kk
parse   : .../std/core/maybe.kk
parse   : .../std/core/either.kk
parse   : .../std/core/debug.kk
parse   : .../std/core/delayed.kk
parse   : .../std/core/show.kk
parse   : .../std/core/tuple.kk
parse   : .../fibonacci.kk
parse   : .../std/core.kk
parse   : .../std/core/console.kk
check   : std/core/types
check   : std/core/unsafe
check   : std/core/undiv
check   : std/core/int
check   : std/core/bool
check   : std/core/hnd
check   : std/core/order
check   : std/core/char
check   : std/core/tuple
check   : std/core/exn
check   : std/core/either
check   : std/core/delayed
check   : std/core/vector
check   : std/core/maybe
check   : std/core/string
check   : std/core/list
check   : std/core/sslice
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_undiv.c:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_undiv.hIn file included from :.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.c:2:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: In file included from warning: .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_unsafe.c:non-void function does not return a value in all control paths [-Wreturn-type]2:

In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_unsafe.h:6:
6In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_bool.c:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h :1702 | }
      | .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:
^.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_int.c
:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_int.h:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
 1702 | }
      | ^
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_bool.h:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
 1702 | }
      | ^
:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
1702:1: warning:  1702 | non-void function does not return a value in all control paths [-Wreturn-type]}

      | ^
 1702 | }
      | ^
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_char.c:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_char.h:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
 1702 | }
      | ^
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.c:15:18: error: call to undeclared function 'kk_cctx_copy'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
   15 |   kk_box_t res = kk_cctx_copy(acc2.res,kk_field_addr_ptr(acc2.holeptr,ctx),&holeptr,ctx);
      |                  ^
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.c:15:12: error: initializing 'kk_box_t' (aka 'struct kk_box_s') with an expression of incompatible type 'int'
   15 |   kk_box_t res = kk_cctx_copy(acc2.res,kk_field_addr_ptr(acc2.holeptr,ctx),&holeptr,ctx);
      |            ^     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_order.c:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_order.h:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
 1702 | }
      | ^
1 warning and 2 errors generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_exn.c:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_exn.h:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
 1702 | }
      | ^
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_string.c:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_string.h:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
 1702 | }
      | ^
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_vector.c:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_vector.h:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
 1702 | }
      | ^
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_tuple.c:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_tuple.h:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
 1702 | }
      | ^
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_maybe.c:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_maybe.h:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
 1702 | }
      | ^
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_delayed.c:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_delayed.h:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
 1702 | }
      | ^
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_hnd.c:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_hnd.h:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
 1702 | }
      | ^
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_either.c:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_either.h:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
 1702 | }
      | ^
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_sslice.c:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_sslice.h:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
 1702 | }
      | ^
check   : std/core/show
check   : std/core/console
check   : std/core/debug
check   : std/core
check   : fibonacci
1 warning generated.
1 warning generated.
1 warning generated.
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/fibonacci.c:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/fibonacci.h:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_show.c:2:
In file included from  .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_show.h1702:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702: | }
      | ^
1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
 1702 | }
      | ^
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_list.c:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_list.h:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
 1702 | }
      | ^
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_debug.c:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_debug.h:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
 1702 | }
      | ^
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_console.c:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_console.h:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
 1702 | }
      | ^
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core.c:2:
In file included from .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core.h:6:
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.h:1702:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
 1702 | }
      | ^
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
std/core/types(1, 1): build error: user error (error  : command failed (exit code 1)
command: /Users/kyle/emsdk/upstream/emscripten/emcc -Wall -Wextra -Wpointer-arith -Wshadow -Wstrict-aliasing -Wno-unknown-pragmas -Wno-missing-field-initializers -Wno-unused-parameter -Wno-unused-variable -Wno-unused-value -Wno-unused-but-set-variable -g -Og -c -D__wasi__ -I /usr/local/share/koka/v3.1.2/kklib/include -o .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.o .koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.c)

Failed to compile fibonacci.kk
```

Here is a smaller excerpt with only the errors:

```txt
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.c:15:18: error: call to undeclared function 'kk_cctx_copy'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
   15 |   kk_box_t res = kk_cctx_copy(acc2.res,kk_field_addr_ptr(acc2.holeptr,ctx),&holeptr,ctx);
      |                  ^
.koka/v3.1.2/emcc-stdalloc-wasm32-debug-56cc25/std_core_types.c:15:12: error: initializing 'kk_box_t' (aka 'struct kk_box_s') with an expression of incompatible type 'int'
   15 |   kk_box_t res = kk_cctx_copy(acc2.res,kk_field_addr_ptr(acc2.holeptr,ctx),&holeptr,ctx);
```

**Comment by daanx** [comment]
Ah, that's because the WASM target does not have a `malloc_usable_size` equivalent (I think?) :-(  Let me look if I can repro as I think we just need to tell mimalloc that we are targeting wasm. 

How did you install emscripten on your mac (I have an M1 too)

**Comment by daanx** [comment]
ps. Can you change the line in `kklib/include/kklib.h`  as in the above commit (a979e1c) ? That should fix the errors you are getting now with `-fstdalloc` (I hope).

**Comment by daanx** [comment]
I just tried on an M1 -- I had already emscripten and wasmtime installed with `brew`; after `brew upgrade wasmtime emscripten`  the versions are `24.0.0` for wasmtime and `3.1.67` for emscripten (!) where `emcc --version` says 
```
shared:INFO: (Emscripten: Running sanity checks)
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.67-git
Copyright (C) 2014 the Emscripten authors (see AUTHORS.txt)
This is free and open source software under the MIT license.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

It all works -- but on the latest Koka build from the `dev` branch. Koka v3.1.1 works too, but v3.1.2 gives the same error as you got -- I think that is just an older version of the `mimalloc` library. I tried the latest `dev` of Koka with `-fstdalloc` (with the above fix) and that worked as well.
